### PR TITLE
User report cleanup

### PR DESCRIPTION
**Pull request by mikehelmick**
## Proposed Changes

* Fix user report cleanup

**Release Note**

```release-note
Fix a potential race condition where the system could be configured to clean up user report records (delete nonce) while the user report code is still valid. This ensures that user reports across all realms are not cleaned up sooner than the realm with the longest short code timeout AND user report enabled.
```

### CODE REVIEW

File: pkg/controller/cleanup/handle_cleanup.go [hunk 1]
```diff
@@ -266,7 +267,27 @@ func (c *Controller) HandleCleanup() http.Handler {
 		func() {
 			defer enobs.RecordLatency(ctx, time.Now(), mLatencyMs, &result, &item)
 			item = tag.Upsert(itemTagKey, "UNCLAIMED_USER_REPORTS")
-			if count, err := c.db.PurgeUnclaimedUserReports(c.config.UserReportUnclaimedMaxAge); err != nil {
+
+			realms, _, err := c.db.ListRealms(pagination.UnlimitedResults)
```
Comment (sethvargo) [inline pkg/controller/cleanup/handle_cleanup.go:271]: Instead of looping over all the realms, can we just `SELECT MAX(user_report_unclaimed_max_age)`?

Comment (sethvargo) [reply pkg/controller/cleanup/handle_cleanup.go:271]: Sorry, I meant `MAX(realm.short_code_max_minutes)`

Comment (mikehelmick) [reply pkg/controller/cleanup/handle_cleanup.go]: updated

File: pkg/database/realm.go [hunk 2]
```diff
@@ -1294,6 +1294,25 @@ func (r *Realm) ValidTestType(typ string) bool {
 	}
 }
 
+func (db *Database) MaximumUserReportTimeout() (time.Duration, error) {
+	// TestTypeUSerReport == 16 == 0b10000
```
Comment (sethvargo) [inline pkg/database/realm.go:1298]:
```suggestion
	// TestTypeUserReport == 16 == 0b10000
```

File: pkg/database/realm.go [hunk 3]
```diff
@@ -1294,6 +1294,25 @@ func (r *Realm) ValidTestType(typ string) bool {
...
+	sql := `
+	SELECT
+		MAX(code_duration) AS duration
+	FROM realms
+	WHERE
+	    allowed_test_types >= 16;`
+
+	var result struct {
+		Quantity int64 `gorm:"column:duration;"`
```
Comment (sethvargo) [inline pkg/database/realm.go:1307]: Quantity?

**Review by sethvargo — APPROVED** [review]

File: pkg/database/realm.go (hunk 3 above)
Comment (mikehelmick) [reply pkg/database/realm.go:1307]: copy paste...

**Pull request merged**